lib/remoteparticipant.js

```diff
@@ -44,49 +54,84 @@ class RemoteParticipant extends Participant {
   /**
    * @private
    */
-  _addTrack(remoteTrack) {
+  _addTrack(remoteTrack, publication) {
     if (!super._addTrack.call(this, remoteTrack)) {
       return null;
     }
+    if (publication) {
+      remoteTrack.setSid(publication.trackSid);
+      publication.subscribed(remoteTrack);
+    }
     this.emit('trackSubscribed', remoteTrack);
     return remoteTrack;
   }
 
   /**
    * @private
    */
-  _unsubscribeTracks() {
-    const tracks = Array.from(this.tracks.values());
-    tracks.forEach(this._unsubscribeTrack, this);
+  _addTrackPublication(publication) {
+    const addedPublication = super._addTrackPublication(publication);
+    if (!addedPublication) {
+      return null;
+    }
+    this.emit('trackPublished', addedPublication);
+    return addedPublication;
+  }
+  /**
+   * @private
+   */
+  _getTrackPublicationEvents() {
+    return [
+      ...super._getTrackPublicationEvents(),
+      ['subscriptionFailed', 'trackSubscriptionFailed']
+    ];
   }
 
   /**
    * @private
    */
-  _unsubscribeTrack(remoteTrack) {
-    const unsubscribedTrack = this.tracks.get(remoteTrack.id);
-    if (unsubscribedTrack) {
-      unsubscribedTrack._unsubscribe();
-      this.emit('trackUnsubscribed', unsubscribedTrack);
-    }
+  _unsubscribeTracks() {
+    this.trackPublications.forEach(publication => {
+      const track = publication.track;
+      if (track) {
+        track.unsubscribe();
+      }
+      publication.unsubscribe();
```

Comment:
Should `publication` be responsible for calling `track.unsubscribe`? Seems like that is more of the Publication's concern than the RemoteParticipant's.